packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,462 @@
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key type to extract from.
+ * @template Name - The name whose key to extract.
+ */
+export type ExtractPermissionTargetKey<
+  Key extends string,
+  Name extends string,
+> = Key extends Name ? Key : Extract<Key, KeyOfTargetName<Key, Name>>;
+
+/**
+ * Internal utility for extracting the members types of an array. The type
+ * evalutes to `never` if the specified type is the empty tuple or neither
+ * an array nor a tuple.
+ *
+ * @template ArrayType - The array type whose members to extract.
+ */
+type ExtractArrayMembers<ArrayType> = ArrayType extends
+  | readonly [...infer ArrayMembers]
+  | [...infer ArrayMembers]
+  ? ArrayMembers[number]
+  : never;
+
+/**
+ * A utility type for extracting the allowed caveat types for a particular
+ * permission from a permission specification type or type union.
+ *
+ * @template PermissionSpecification - The permission specification type to
+ * extract valid caveat types from.
+ * @template TargetName - The target name of the permission whose allowed
+ * caveats to extract.
+ */
+export type ExtractAllowedCaveatTypes<
+  PermissionSpecification extends PermissionSpecificationConstraint,
+  TargetName extends string,
+> = PermissionSpecification extends PermissionSpecificationConstraint & {
+  targetKey: ExtractPermissionTargetKey<
+    PermissionSpecification['targetKey'],
+    TargetName
+  >;
+}
+  ? ExtractArrayMembers<PermissionSpecification['allowedCaveats']>
+  : never;
+
+/**
+ * The options object of {@link constructPermission}.
+ *
+ * @template TargetPermission - The {@link Permission} that will be constructed.
+ */
+export type PermissionOptions<TargetPermission extends PermissionConstraint> = {
+  target: TargetPermission['parentCapability'];
+  /**
+   * The origin string of the subject that has the permission.
+   */
+  invoker: OriginString;
+
+  /**
+   * The GUID of the permission object.
+   * Assigned if not provided.
+   */
+  id?: string;
+
+  /**
+   * The caveats of the permission.
+   * See {@link Caveat}.
+   */
+  caveats?: NonEmptyArray<CaveatConstraint>;
+  // TODO: Refactor allowed caveats validation to enable this?
+  // caveats?: NonEmptyArray<ExtractCaveats<TargetPermission>>;
+};
+
+/**
+ * The default permission factory function. Naively constructs a permission from
+ * the inputs. Sets a default, random `id` if none is provided.
+ *
+ * @see {@link Permission} For more details.
+ *
+ * @template TargetPermission - The {@link Permission} that will be constructed.
+ * @param options - The options for the permission.
+ * @returns The new permission object.
+ */
+export function constructPermission<
+  TargetPermission extends PermissionConstraint,
+>(options: PermissionOptions<TargetPermission>): PermissionConstraint {
+  const { caveats = null, id, invoker, target } = options;
+
+  return {
+    id: id ?? nanoid(),
```

Comment:
Is this `id` ever expected to be passed in?

Reply:
It isn't at the moment, but 🤷‍♂️

Reply:
I think it would be simpler to always use `nanoid` here then, and eliminate this branch. For the same reasons as given here: https://github.com/MetaMask/snaps-skunkworks/pull/28#discussion_r746991911

Reply:
118bbc9